tracing.go

```diff
@@ -84,53 +87,155 @@ func (s Span) ParentID() uint64 { return s.parentID }
+	ext.PeerService.Set(span, serviceName)
+	setPeerHostPort(span, c.remotePeerInfo.HostPort)
+	span.SetTag("as", call.callReq.Headers[ArgScheme])
+	call.callReq.Tracing.initFromOpenTracing(span)
+	return span
 }
 
-func (s *Span) sampleRootSpan(sampleRate float64) {
-	// Sampling only affects root spans
-	if s.ParentID() != 0 {
-		return
+// InjectOutboundSpan retrieves OpenTracing Span from `response`, where it is stored
+// when the outbound call is initiated. The tracing API is used to serialize the span
+// into the application `headers`, which will propagate tracing context to the server.
+// Returns modified headers containing serialized tracing context.
+func InjectOutboundSpan(response *OutboundCallResponse, headers map[string]string) map[string]string {
+	span := response.span
+	if span == nil {
+		return headers
+	}
+	if headers == nil {
+		headers = make(map[string]string)
+	}
+	carrier := opentracing.TextMapCarrier(headers)
+	if err := span.Tracer().Inject(span.Context(), opentracing.TextMap, carrier); err == nil {
+		return headers
 	}
+	return headers
+}
 
-	if rand.Float64() > sampleRate {
-		s.EnableTracing(false)
+// extractInboundSpan attempts to create a new OpenTracing Span for inbound request
+// using only trace IDs stored in the frame's tracing field. It only works if the
+// tracer understand Zipkin-style trace IDs. If such attempt fails, another attempt
+// will be made from the higher level function ExtractInboundSpan() once the
+// application headers are read from the wire.
+func (c *Connection) extractInboundSpan(callReq *callReq) opentracing.Span {
+	spanCtx, err := c.Tracer().Extract(ZipkinSpanFormat, &callReq.Tracing)
+	if spanCtx != nil {
+		span := c.Tracer().StartSpan("", ext.RPCServerOption(spanCtx))
+		ext.PeerService.Set(span, callReq.Headers[CallerName])
+		span.SetTag("as", callReq.Headers[ArgScheme])
+		setPeerHostPort(span, c.remotePeerInfo.HostPort)
+		return span
+	}
+	if err != opentracing.ErrUnsupportedFormat && err != opentracing.ErrSpanContextNotFound {
+		c.log.Error("Failed to extract Zipkin-style span: " + err.Error())
 	}
+	return nil
 }
 
-func newSpan(traceID, spanID, parentID uint64, tracingEnabled bool) *Span {
-	flags := byte(0)
-	if tracingEnabled {
-		flags = tracingFlagEnabled
+// ExtractInboundSpan is a higher level version of extractInboundSpan().
+// If the lower-level attempt to create a span from incoming request was
+// successful (e.g. when then Tracer supports Zipkin-style trace IDs),
+// then the application headers are only used to read the Baggage and add
+// it to the existing span. Otherwise, the standard OpenTracing API supported
+// by all tracers is used to deserialize the tracing context from the
+// application headers and start a new server-side span.
+// Once the span is started, it is wrapped in a new Context, which is returned.
+func ExtractInboundSpan(ctx context.Context, call *InboundCall, headers map[string]string, tracer opentracing.Tracer) context.Context {
```

Comment:
>  but why do you think that the tracing middleware should infect user-owned application headers instead of being transparent

I am not against it being transparent, my point is that it's not the business of OpenTracing API, which is transport-agnostic, to deal with hiding the headers, it's the business of the RPC framework that's using OpenTracing API to instrument itself (i.e. _this_ change). In fact, OpenTracing API supports a binary format for tracing context, so that we could transmit the whole thing (including baggage) as a single app header with the name known only to tchannel (e.g. `tchannel-opentracing-context={base64blob}`), and it can be easily removed from the app headers on the receiving side. But if we want to use "text map" format of encoding tracing context (as we do in these PRs, to make them readable on the wire), OpenTracing API returns it as a map, and the only way to distinguish them on the receiving side is by prefixing each key in that map with tchannel-specific header. This will require at least one extra map allocations (keys can be cached).